File: frontend/admin/js/reports.js
/* ═══════════════ REPORTS ═══════════════
   Charts are inline SVG built from the same numbers shown in the table
   beside them — no chart library, and nothing that can render a shape the
   figures don't support. Every screen says what it counted and what it
   excluded, because a total that quietly includes cancellations is the
   classic way a dashboard lies. */

const repState = {days:30};

/* Minutes → the unit a human would use. A stage cleared in four minutes
   shown as "0 h" reads as broken instrumentation rather than as speed. */
function dur(mins){
  if(mins === null || mins === undefined) return '—';
  if(mins < 1)   return '< 1 min';
  if(mins < 90)  return Math.round(mins) + ' min';
  if(mins < 2880) return (mins/60).toFixed(1) + ' h';
  return (mins/1440).toFixed(1) + ' days';
}

async function renderReports(el, sub){
  const tabs = [['sales','Sales'],['production-throughput','Throughput'],
                ['stock-valuation','Stock valuation'],['ai-usage','AI usage']];
  el.innerHTML = `<div class="row" style="margin-bottom:18px">
    ${tabs.map(([k,label])=>`<button class="btn btn-sm ${(sub||'sales')===k?'btn-primary':'btn-quiet'}"
      onclick="goTo('reports','${k}')">${esc(label)}</button>`).join('')}
  </div><div id="repBody"></div>`;
  const body = document.getElementById('repBody');
  if(sub === 'production-throughput') return renderThroughput(body);
  if(sub === 'stock-valuation')       return renderStockValue(body);
  if(sub === 'ai-usage')              return renderAiUsage(body);
  return renderSales(body);
}

/* Bar chart as inline SVG. viewBox + preserveAspectRatio="none" lets it fill
   whatever width the card has without measuring anything in JS. */
function barChart(points, valueKey, label){
  const max = Math.max(1, ...points.map(p=>p[valueKey]));
  const n = points.length, w = 100 / n;
  const bars = points.map((p,i)=>{
    const h = (p[valueKey] / max) * 100;
    return `<rect x="${(i*w).toFixed(3)}" y="${(100-h).toFixed(3)}"
      width="${(w*0.72).toFixed(3)}" height="${h.toFixed(3)}"
      ><title>${esc(p.date)} · ${esc(String(p[valueKey]))}</title></rect>`;
  }).join('');
  const first = points[0], last = points[points.length-1];
  return `<div class="chart">
    <svg viewBox="0 0 100 100" preserveAspectRatio="none" role="img"
      aria-label="${esc(label)}">${bars}</svg>
    <div class="chart-axis">
      <span>${esc(first?fmtDate(first.date):'')}</span>
      <span class="dim">peak ${esc(String(max))}</span>
      <span>${esc(last?fmtDate(last.date):'')}</span>
    </div>
  </div>`;
}

/* ── Sales ───────────────────────────────────────────────────── */
async function renderSales(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/reports/sales?days='+repState.days);
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  const t = d.totals;
  const gstTotal = d.gst.reduce((a,g)=>a+g.tax,0);
  el.innerHTML = `
    <div class="row" style="margin-bottom:14px">
      ${[30,90,365].map(n=>`<button class="chip ${repState.days===n?'on':''}"
        onclick="repState.days=${n};renderSales(document.getElementById('repBody'))">
        ${n===365?'1 year':n+' days'}</button>`).join('')}
    </div>
    <div class="grid g3" style="margin-bottom:18px">
      <div class="stat"><span>Revenue</span><b>${money(t.revenue)}</b></div>
      <div class="stat"><span>Orders</span><b>${t.orders}</b></div>
      <div class="stat"><span>Average order</span><b>${money(t.aov)}</b></div>
    </div>
    <div class="card">
      <h2>Revenue per day</h2>
      <p class="tiny muted">${t.pieces} piece(s) across ${t.orders} order(s).
        ${t.cancelled} cancelled order(s) are counted below but contribute no revenue.</p>
      ${barChart(d.series,'revenue','Revenue per day')}
    </div>
    <div class="card">
      <h2>By product</h2>
      ${d.products.length ? `<table>
        <thead><tr><th>Product</th><th>Pieces</th><th>Lines</th><th>Revenue</th></tr></thead>
        <tbody>${d.products.map(p=>`<tr>
          <td data-label="Product">${esc(p.product)}</td>
          <td data-label="Pieces">${p.qty}</td>
          <td data-label="Lines">${p.orders}</td>
          <td data-label="Revenue">${money(p.revenue)}</td>
        </tr>`).join('')}</tbody></table>` : '<div class="empty">Nothing sold in this window.</div>'}
    </div>
    <div class="card">
      <h2>GST collected</h2>
      ${d.gst.length ? `<table>
        <thead><tr><th>Rate</th><th>Taxable value</th><th>Tax</th></tr></thead>
        <tbody>${d.gst.map(g=>`<tr>
          <td data-label="Rate">${g.rate}%</td>
          <td data-label="Taxable">${money(g.taxable)}</td>
          <td data-label="Tax">${money(g.tax)}</td>
        </tr>`).join('')}
        <tr><td><b>Total</b></td><td></td><td><b>${money(gstTotal)}</b></td></tr>
        </tbody></table>` : '<div class="empty">No tax breakdowns in this window.</div>'}
      ${d.gst_covered < d.gst_total_orders ? `<p class="tiny dim" style="margin-top:10px">
        Covers ${d.gst_covered} of ${d.gst_total_orders} orders — the rest were
        placed before per-order tax breakdowns were stored, so their GST can't
        be split by rate here. Their revenue is still in the totals above.</p>`:''}
      <p class="tiny dim" style="margin-top:8px">This is what was charged, not
        a return. Reconcile against your filings before submitting anything.</p>
    </div>`;
}

/* ── Throughput ──────────────────────────────────────────────── */
async function renderThroughput(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/reports/throughput');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  const inFlight = d.pipeline.reduce((a,p)=>a+(p.stage<5?p.orders:0),0);
  el.innerHTML = `
    <div class="grid g3" style="margin-bottom:18px">
      <div class="stat"><span>In the pipeline</span><b>${inFlight}</b></div>
      <div class="stat"><span>Delivered</span><b>${(d.pipeline[5]||{}).orders||0}</b></div>
      <div class="stat"><span>Proofs approved</span><b>${d.proofs_approved}</b></div>
    </div>
    <div class="card">
      <h2>Where orders are sitting</h2>
      <table><tbody>${d.pipeline.map(p=>`<tr>
        <td data-label="Stage">${esc(p.label)}</td>
        <td data-label="Orders"><b>${p.orders}</b></td>
        <td data-label="" style="width:55%">
          <div class="meter"><i style="width:${inFlight+((d.pipeline[5]||{}).orders||0)
            ? (p.orders/Math.max(1,...d.pipeline.map(x=>x.orders))*100).toFixed(1) : 0}%"></i></div></td>
      </tr>`).join('')}</tbody></table>
      <p class="tiny dim">Cancelled orders excluded.</p>
    </div>
    <div class="card">
      <h2>Time spent at each stage</h2>
      <p class="tiny muted">From the audit log, counting only stages an order
        has actually left. An order still sitting somewhere has no duration
        yet — assuming "until now" would make a stalled queue look fast.</p>
      <table>
        <thead><tr><th>Stage</th><th>Average</th><th>Measured from</th></tr></thead>
        <tbody>${d.stage_time.map(s=>`<tr>
          <td data-label="Stage">${esc(s.label)}</td>
          <td data-label="Average">${s.avg_minutes===null?'<span class="dim">—</span>'
            :'<b>'+dur(s.avg_minutes)+'</b>'}</td>
          <td data-label="Sample" class="tiny dim">${s.n} completed move(s)</td>
        </tr>`).join('')}</tbody></table>
    </div>
    <div class="card">
      <h2>Dispatched per day</h2>
      ${barChart(d.dispatched,'count','Parcels dispatched per day')}
    </div>`;
}

/* ── Stock valuation ─────────────────────────────────────────── */
async function renderStockValue(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/reports/stock');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  el.innerHTML = `
    ${d.missing_cost ? `<div class="alert warn" style="margin-bottom:16px">
      <span class="material-symbols-outlined">warning</span>
      <span><b>${d.missing_cost} product(s) have no cost price.</b> They value at
        zero, so the figure below is not what your stock is worth. Set cost
        prices under Inventory → the product editor.</span>
    </div>`:''}
    <div class="grid g3" style="margin-bottom:18px">
      <div class="stat"><span>At cost</span><b>${money(d.cost_value)}</b></div>
      <div class="stat"><span>At lowest sell price</span><b>${money(d.retail_value)}</b></div>
      <div class="stat"><span>Variants counted</span><b>${d.counted_variants}/${d.total_variants}</b></div>
    </div>
    <div class="card" style="padding:0;overflow:hidden">
      <table>
        <thead><tr><th>Product</th><th>On hand</th><th>Unit cost</th>
          <th>At cost</th><th>At sell</th></tr></thead>
        <tbody>${d.products.map(p=>`<tr class="${p.negative?'row-off':''}">
          <td data-label="Product">${esc(p.product)}
            ${p.negative?`<br><span class="badge badge-warn">${p.negative} negative</span>`:''}</td>
          <td data-label="On hand">${p.qty}</td>
          <td data-label="Unit cost">${p.cost?money(p.cost):'<span class="dim">not set</span>'}</td>
          <td data-label="At cost">${money(p.value)}</td>
          <td data-label="At sell">${money(p.retail)}</td>
        </tr>`).join('')}</tbody>
      </table>
    </div>
    <p class="tiny dim" style="margin-top:12px">
      Only ${d.counted_variants} of ${d.total_variants} variants have ever had a
      stock movement — the rest have never been counted in, so they read as
      zero rather than as genuinely empty.
    </p>`;
}

/* ── AI usage ────────────────────────────────────────────────── */
async function renderAiUsage(el){
  el.innerHTML = '<div class="empty">Loading…</div>';
  const d = await api('/api/admin/reports/ai');
  if(!d.ok){ el.innerHTML = `<div class="empty">${esc(d.error)}</div>`; return; }
  el.innerHTML = `
    <div class="grid g3" style="margin-bottom:18px">
      <div class="stat"><span>Generations, all time</span><b>${d.total.count}</b></div>
      <div class="stat"><span>Spend, all time</span><b>${money(d.total.cost)}</b></div>
      <div class="stat"><span>Last 30 days</span><b>${d.series.reduce((a,s)=>a+s.count,0)}</b></div>
    </div>
    <div class="card">
      <h2>Generations per day</h2>
      ${barChart(d.series,'count','AI generations per day')}
    </div>
    <div class="card">
      <h2>By model</h2>
      ${d.models.length ? `<table>
        <thead><tr><th>Model</th><th>Generations</th><th>Cost</th></tr></thead>
        <tbody>${d.models.map(m=>`<tr>
          <td data-label="Model">${esc(m.model)}</td>
          <td data-label="Generations">${m.count}</td>
          <td data-label="Cost">${m.cost?money(m.cost):'<span class="dim">free</span>'}</td>
        </tr>`).join('')}</tbody></table>` : '<div class="empty">Nothing generated yet.</div>'}
      <p class="tiny dim" style="margin-top:10px">Pollinations and Gemini are
        free tiers and log a zero cost — a low total here means the free
        providers are carrying the load, not that nothing was generated.</p>
    </div>`;
}

File: backend/reports.py
# ── Reports blueprint ────────────────────────────────────────────
#
# Every figure here is computed from the same rows the rest of the panel
# reads — nothing is precomputed into a summary table, because a summary
# that drifts from the orders it claims to describe is worse than no report.
# At this size the aggregates are milliseconds.
#
# Cancelled orders are excluded from money and included in counts, and each
# report says which it did. "Revenue" that quietly includes cancellations is
# the classic way a dashboard lies.
import json
from datetime import datetime, timedelta, timezone

from flask import Blueprint, request, jsonify

from db import get_db
from permissions import require

reports_bp = Blueprint("reports", __name__, url_prefix="/api/admin/reports")

STAGE_LABELS = ["Proof sent", "Approved", "Printing", "Quality check",
                "Shipped", "Delivered"]


def _days(n):
    """Ordered list of the last n dates as YYYY-MM-DD, oldest first. Built in
    Python rather than SQL so days with no orders still appear — a gap in a
    chart is information, a missing row is just absence."""
    today = datetime.now(timezone.utc).date()
    return [(today - timedelta(days=i)).isoformat() for i in range(n - 1, -1, -1)]


@reports_bp.route("/sales")
@require("reports")
def sales():
    db = get_db()
    try:
        span = max(7, min(365, int(request.args.get("days") or 30)))
    except ValueError:
        span = 30
    days = _days(span)
    since = days[0]

    rows = db.execute(
        """SELECT date(created) d, COUNT(*) n,
                  SUM(CASE WHEN cancelled=0 THEN total_inr ELSE 0 END) rev,
                  SUM(CASE WHEN cancelled=1 THEN 1 ELSE 0 END) cancelled
           FROM orders WHERE date(created) >= date(?) AND payment_status!='pending'
           GROUP BY date(created)""", (since,)).fetchall()
    by_day = {r["d"]: r for r in rows}
    series = [{"date": d,
               "orders": by_day[d]["n"] if d in by_day else 0,
               "revenue": (by_day[d]["rev"] or 0) if d in by_day else 0,
               "cancelled": by_day[d]["cancelled"] if d in by_day else 0}
              for d in days]

    # Per product and the tax split — walked in Python because the figures
    # live inside items_json and tax_json, not in columns we can GROUP BY.
    prod, gst_by_rate, pieces, taxable = {}, {}, 0, 0.0
    orders = db.execute(
        """SELECT items_json, tax_json, total_inr, cancelled
           FROM orders WHERE date(created) >= date(?) AND cancelled=0 AND payment_status!='pending'""",
        (since,)).fetchall()
    for o in orders:
        try:
            items = json.loads(o["items_json"]) or []
        except (ValueError, TypeError):
            items = []
        for it in items:
            if not isinstance(it, dict):
                continue
            k = it.get("product") or it.get("pid") or "—"
            e = prod.setdefault(k, {"product": k, "qty": 0, "revenue": 0.0, "orders": 0})
            e["qty"] += int(it.get("qty") or 0)
            e["revenue"] += float(it.get("total") or 0)
            e["orders"] += 1
            pieces += int(it.get("qty") or 0)
        if o["tax_json"]:
            try:
                t = json.loads(o["tax_json"])
            except ValueError:
                continue
            taxable += float(t.get("subtotal") or 0)
            for l in t.get("lines") or []:
                rate = l.get("gst_percent")
                if rate is None:
                    continue
                b = gst_by_rate.setdefault(str(rate), {"rate": rate, "taxable": 0.0, "tax": 0.0})
                b["taxable"] += float(l.get("total") or 0)
                b["tax"] += float(l.get("total") or 0) * float(rate) / 100.0

    live = [o for o in orders]
    revenue = sum(o["total_inr"] for o in live)
    return jsonify(ok=True, days=span, series=series,
                   totals={
                       "orders": len(live),
                       "revenue": round(revenue, 2),
                       "aov": round(revenue / len(live), 2) if live else 0,
                       "pieces": pieces,
                       "cancelled": sum(s["cancelled"] for s in series),
                   },
                   products=sorted(prod.values(), key=lambda x: -x["revenue"]),
                   gst=[{**v, "taxable": round(v["taxable"], 2), "tax": round(v["tax"], 2)}
                        for v in sorted(gst_by_rate.values(), key=lambda x: x["rate"])],
                   # Orders placed before tax_json existed carry no breakdown,
                   # so the GST table below can cover fewer orders than the
                   # revenue above. Saying so beats a silent mismatch.
                   gst_covered=sum(1 for o in orders if o["tax_json"]),
                   gst_total_orders=len(orders))


@reports_bp.route("/throughput")
@require("reports")
def throughput():
    db = get_db()
    days = _days(30)

    stages = db.execute(
        """SELECT status, COUNT(*) n FROM orders WHERE cancelled=0 AND payment_status!='pending' GROUP BY status""").fetchall()
    by_stage = {r["status"]: r["n"] for r in stages}
    pipeline = [{"stage": i, "label": STAGE_LABELS[i], "orders": by_stage.get(i, 0)}
                for i in range(len(STAGE_LABELS))]

    # Dispatches per day, from the shipments table — the only record of goods
    # physically leaving, which is what throughput means.
    ship = db.execute(
        """SELECT date(created) d, COUNT(*) n, SUM(boxes) boxes
           FROM shipments WHERE date(created) >= date(?) GROUP BY date(created)""",
        (days[0],)).fetchall()
    sm = {r["d"]: r for r in ship}
    dispatched = [{"date": d, "count": sm[d]["n"] if d in sm else 0,
                   "boxes": (sm[d]["boxes"] or 0) if d in sm else 0} for d in days]

    # How long orders sit at each stage, from the audit log's stage moves.
    # Only pairs where we saw both the entry and the exit — an order still
    # sitting somewhere has no duration yet, and assuming "now" would make
    # a stalled queue look fast.
    moves = db.execute(
        """SELECT entity_id, from_val, to_val, created FROM audit_log
           WHERE entity_type='order' AND action='stage' ORDER BY entity_id, id""").fetchall()
    dwell, last = {}, {}
    for m in moves:
        oid, frm = m["entity_id"], m["from_val"]
        prev = last.get(oid)
        if prev and frm is not None:
            try:
                a = datetime.fromisoformat(prev[1].replace(" ", "T"))
                b = datetime.fromisoformat(m["created"].replace(" ", "T"))
                hrs = (b - a).total_seconds() / 3600.0
                if 0 <= hrs < 24 * 90:
                    d = dwell.setdefault(prev[0], [])
                    d.append(hrs)
            except (ValueError, AttributeError):
                pass
        last[oid] = (m["to_val"], m["created"])
    stage_time = []
    for i, label in enumerate(STAGE_LABELS):
        vals = dwell.get(str(i), [])
        # Kept in minutes: a stage cleared in four minutes rounds to "0.1 h",
        # which reads like a bug rather than like speed. The panel picks the
        # unit; rounding to hours here would throw the detail away first.
        stage_time.append({"stage": i, "label": label, "n": len(vals),
                           "avg_minutes": round(sum(vals) / len(vals) * 60, 1) if vals else None})

    proofed = db.execute(
        """SELECT COUNT(*) c FROM audit_log
           WHERE entity_type='order' AND action='stage'
             AND from_val='0' AND to_val='1'""").fetchone()["c"]
    return jsonify(ok=True, pipeline=pipeline, dispatched=dispatched,
                   stage_time=stage_time, proofs_approved=proofed)


@reports_bp.route("/stock")
@require("reports")
def stock_value():
    db = get_db()
    rows = db.execute(
        """SELECT p.id, p.slug, p.name, p.cost_price,
                  (SELECT COALESCE(MIN(unit_price),p.base_price) FROM price_tiers t
                    WHERE t.product_id=p.id) AS floor_price,
                  COALESCE(SUM(v.stock_qty),0) AS qty,
                  SUM(CASE WHEN v.stock_qty < 0 THEN 1 ELSE 0 END) AS negative
           FROM products p LEFT JOIN variants v ON v.product_id=p.id
           GROUP BY p.id ORDER BY p.sort,p.id""").fetchall()
    out, cost_total, retail_total, no_cost = [], 0.0, 0.0, 0
    for r in rows:
        qty = r["qty"] or 0
        cost = r["cost_price"] or 0
        if not cost:
            no_cost += 1
        cost_total += qty * cost
        retail_total += qty * (r["floor_price"] or 0)
        out.append({"product": r["name"], "product_id": r["slug"], "qty": qty,
                    "cost": cost, "value": round(qty * cost, 2),
                    "retail": round(qty * (r["floor_price"] or 0), 2),
                    "negative": r["negative"] or 0})
    counted = db.execute(
        """SELECT COUNT(DISTINCT variant_id) c FROM stock_moves""").fetchone()["c"]
    variants = db.execute("SELECT COUNT(*) c FROM variants").fetchone()["c"]
    return jsonify(ok=True, products=out,
                   cost_value=round(cost_total, 2), retail_value=round(retail_total, 2),
                   # A valuation is only as good as its cost prices; if none
                   # are set the number is zero and must say why.
                   missing_cost=no_cost,
                   counted_variants=counted, total_variants=variants)


@reports_bp.route("/ai")
@require("reports")
def ai_usage():
    db = get_db()
    days = _days(30)
    rows = db.execute(
        """SELECT date(created) d, COUNT(*) n, COALESCE(SUM(cost_inr),0) cost
           FROM generations WHERE date(created) >= date(?) GROUP BY date(created)""",
        (days[0],)).fetchall()
    m = {r["d"]: r for r in rows}
    series = [{"date": d, "count": m[d]["n"] if d in m else 0,
               "cost": round(m[d]["cost"], 2) if d in m else 0} for d in days]
    models = db.execute(
        """SELECT COALESCE(model,'unknown') model, COUNT(*) n,
                  COALESCE(SUM(cost_inr),0) cost
           FROM generations GROUP BY model ORDER BY n DESC""").fetchall()
    allt = db.execute(
        "SELECT COUNT(*) n, COALESCE(SUM(cost_inr),0) cost FROM generations").fetchone()
    return jsonify(ok=True, series=series,
                   models=[{"model": r["model"], "count": r["n"],
                            "cost": round(r["cost"], 2)} for r in models],
                   total={"count": allt["n"], "cost": round(allt["cost"], 2)})
